Document guidelines for physical operator yielding

carols10cents opened the pull request:
To start a policy of the behavior physical operator streams should have and drive improvements in this area to allow for timely cancellation.

I'm very open to any wording changes and looking forward to discussion on whether I've accurately captured the intent of datafusion!

## Which issue does this PR close?

Connects to #14036 and related to pull requests such as #14028.

## Rationale for this change

Datafusion should be explicit about its aspirations for what it provides as well as provide guidelines for trait implementers about yielding to enable timely cancellation.

## What changes are included in this PR?

A policy statement in the documentation with some reference links.

## Are these changes tested?

No, this is only documentation.

## Are there any user-facing changes?

This is the documentation :)

alamb commented on datafusion/physical-plan/src/execution_plan.rs:
```diff
@@ -260,13 +260,30 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
+    /// the overhead-heavy syscall needed to take the time) or by counting rows or
+    /// batches.
+    ///
+    /// The goal is for `datafusion`-provided operator implementation to
+    /// strive for [the guideline of not spending a long time without reaching
+    /// an `await`/yield point][async-guideline]. Progress towards this goal
+    /// is tracked partially by the cancellation benchmark.
```
I recommend making this the sentence higher up as the rationale is important

```
pin the CPU and must yield back to the tokio runtime regularly. ... <move this paragraph up there>
```

I think we could potentially make the wording more concise, something like

```suggestion
    /// `ExecutionPlan` implementations should follow
    /// [the guideline of not spending a long time without reaching
    /// an `await`/yield point][async-guideline]. 
```

In terms of the cancellation benchmark, I would suggest a sentence with a link to the actual benchmark -- https://github.com/apache/datafusion/blob/main/benchmarks/README.md#benchmarks

Actually, now that I went to look it seems like we should docment the cancellation benchmark there too 🤔

alamb reviewed (commented):
Thank you so much @carols10cents  -- I think this is super valuable, and in my opinion is not changing the DataFusion semantics, but simply documenting what is implicitly already done in the code.

cc @berkaysynnada / @ozankabak  in case you have other thoughts in this area.

berkaysynnada commented on datafusion/physical-plan/src/execution_plan.rs:
```diff
@@ -260,13 +260,30 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
+    /// batches.
+    ///
+    /// The goal is for `datafusion`-provided operator implementation to
+    /// strive for [the guideline of not spending a long time without reaching
```
I don't know if there is any implementation such that it manually yields because of spending long time without a yield point (for CPU bound works), and I am also not very sure we could need that.

carols10cents replied:
I'm not sure if I'm understanding what you're saying correctly-- but I believe datafusion _does_ need this as demonstrated by [the cancellation benchmarks I recently added](https://github.com/apache/datafusion/pull/14818) that show there's at least one case where it takes 32ms (on @alamb's machine) to cancel/drop the runtime because there are operations that aren't yielding often enough (the guidelines I've seen suggest aiming for 1ms of work between yield points).

There are also issues such as https://github.com/apache/datafusion/issues/14036 where queries aren't able to be cancelled, and the root cause also appears to be not yielding often enough.

Could you elaborate on why you _don't_ think datafusion needs manual yields, based on the behavior of the benchmarks and uncancellable query issues?

ozankabak replied:
I think what he means is that in most cases a manual yield shouldn't be necessary for CPU-bound operators in most cases (not all). It is possible for this to be necessary in certain situations (very large batch sizes, operator does superlinear-complexity work w.r.t. batch size etc.), but it shouldn't be a common situation. I think the documentation should probably state this, and give some concrete examples when manual yielding may be necessary.

crepererum replied:
> I think what he means is that in most cases a manual yield shouldn't be necessary for CPU-bound operators in most cases (not all).

I think we can state that if you work packages are record batches and your compute complexity is linear (like filter and map/project operation), then you probably don't need this. But if you do any form of aggregation or super-linear behavior (e.g. unnest, data decompression), then you must think about that issue.

carols10cents replied:
Could someone make a concrete suggestion here on how the text should be changed for discussion?

carols10cents commented on datafusion/physical-plan/src/execution_plan.rs:
```diff
@@ -260,13 +260,30 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
+    /// To enable timely cancellation, the [`Stream`] that is returned must not
+    /// pin the CPU and must yield back to the tokio runtime regularly. This can
+    /// be achieved by manually returning [`Poll::Pending`] in regular intervals,
+    /// or the use of [`tokio::task::yield_now()`]. Cooperative scheduling may also
```
Ah, this was something @crepererum had discussed -- I'm fine taking it out now as it's not actually public in a released versions of tokio yet.

crepererum replied:
I vote for leaving it in or at least put it back once the relevant tokio version is relased. It's a reasonable API with a reasonable behavior. If you wonder how much you use it, that question is something you also need to answer for every other `Pending` check/counter. So only because it doesn't give you an universal answer doesn't mean it's an inappropriate tool.

carols10cents replied:
Yeah, after thinking about this more I don't think it makes sense to point to something that datafusion isn't even using yet (because it hasn't been released). When the `coop` functionality is released and datafusion uses it, then it makes more sense to document here.

carols10cents commented on datafusion/physical-plan/src/execution_plan.rs:
```diff
@@ -260,13 +260,30 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
     /// Thus, [`spawn`] is disallowed, and instead use [`SpawnedTask`].
     ///
+    /// To enable timely cancellation, the [`Stream`] that is returned must not
+    /// pin the CPU and must yield back to the tokio runtime regularly. This can
```
> What do we mean by "pin the CPU" ?

Doing blocking work in an async task. Do you have a suggestion on different wording that would be clearer?

> Moreover, returning just Poll::Pending and calling yield_now() does not exactly do the same thing.

Right, I didn't mean to imply that they did, I was hoping it would come across as "pick the appropriate one for the particular situation". Would adding an "(as appropriate)" here address your comment, or do you have a different suggestion on how to reword this?

crepererum replied:
> In the current architecture of datafusion, we are only introduced with Pending results because of an IO work. So, tasks can only yield at IO points (there are a few exceptions).

This is not true, we already do cooperative yielding, see #5299 and the bug report that was linked in there.

ozankabak replied:
Right - but that is the only place AFAICT we do this. So it is (and should be) very rare - which the docs ought to mention.

berkaysynnada replied:
> > In the current architecture of datafusion, we are only introduced with Pending results because of an IO work. So, tasks can only yield at IO points (there are a few exceptions).
> 
> This is not true, we already do cooperative yielding, see #5299 and the bug report that was linked in there.

That is one of the exceptions I was referring to. However we know that is a very specific edge case arising from the nature of current `RepartitionExec`. Unless Iam overlooking certain cases [this version](https://github.com/apache/datafusion/pull/14411) should not cause any problem in a scenario mentioned in the issue.

I stepped in here to prevent any misunderstanding that might lead people to add manual yields at arbitrary points. While some well-reasoned designs serving to specific use cases may require this, we should emphasize that it should be done consciously. I understand that adding a few random yields may not have a noticeable impact in practice, but this shouldn't spread throughout the codebase.

carols10cents replied:
Could someone make a concrete suggestion here on how the text should be changed for discussion?

crepererum replied:
I don't think it's the only place. https://github.com/apache/datafusion/pull/14028 gives another example where periodic yielding is a good idea.

carols10cents replied:
@crepererum so are you saying similar documentation needs to be added in another place? That is, are you looking for me to change anything (and if so, what precisely) or was your comment for informational purposes only?

crepererum replied:
No no, this was mostly an answer to "Right - but that is the only place AFAICT we do this." and Co.

ozankabak replied:
Touche. I only checked what was on `main` and not unmerged PRs. BTW the verdict is still out on #14028 whether the approach within is the right solution for the problem (it may well be, or not).

Anyway, I don't think there is a disagreement on how to proceed w.r.t. to the documentation.

ozankabak commented:
Thanks for improving the docs, left my suggestions inline

carols10cents commented:
I just pushed some more commits addressing some comments; there is one TODO commit in there that I will update once https://github.com/apache/datafusion/pull/15054 has been merged in so that I can link to the relevant part of the benchmarks readme that I added in that PR. I will squash all these commits when we're done revising.

ozankabak commented on datafusion/physical-plan/src/execution_plan.rs:
```diff
@@ -260,13 +260,28 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
     /// used.
     /// Thus, [`spawn`] is disallowed, and instead use [`SpawnedTask`].
     ///
+    /// To enable timely cancellation, the [`Stream`] that is returned must not
+    /// block the CPU and must yield back to the tokio runtime regularly.
+    /// [`ExecutionPlan`] implementations should follow [the guideline of not
+    /// spending a long time without reaching an `await`/yield point][async-guideline].
+    /// This can be achieved by manually returning [`Poll::Pending`] in regular
+    /// intervals, or the use of [`tokio::task::yield_now()`] (as appropriate).
+    /// Determination for "regularly" may be made using a timer (being careful
+    /// with the  overhead-heavy syscall needed to take the time) or by
+    /// counting rows or batches.
+    ///
+    /// The cancellation benchmark tracks some cases of how quickly queries can
+    /// be cancelled.
```
```suggestion
    /// To enable timely cancellation, the [`Stream`] that is returned must not
    /// block the CPU indefinitely and must yield back to the tokio runtime regularly.
    /// In a typical [`ExecutionPlan`], this automatically happens unless there are
    /// special circumstances; e.g. when the computational complexity of processing a
    /// batch is superlinear. See this [general guideline][async-guideline] for more context
    /// on this point, which explains why one should avoid spending a long time without
    /// reaching an `await`/yield point in asynchronous runtimes.
    /// This can be achieved by manually returning [`Poll::Pending`] and setting up wakers appropriately, or the use of [`tokio::task::yield_now()`] when appropriate.
    /// In special cases that warrant manual yielding, determination for "regularly" may be
    /// made using a timer (being careful with the  overhead-heavy system call needed to
    /// take the time), or by counting rows or batches.
    ///
    /// The cancellation benchmark tracks some cases of how quickly queries can
    /// be cancelled.
```

ozankabak reviewed (commented):
Left a suggestion per your request

alamb commented:
> I just pushed some more commits addressing some comments; there is one TODO commit in there that I will update once #15054 has been merged in so that I can link to the relevant part of the benchmarks readme that I added in that PR. I will squash all these commits when we're done revising.

#15054  has been merged. @carols10cents are you willing to update this PR again? If not I can do so too

carols10cents commented:
> #15054 has been merged. @carols10cents are you willing to update this PR again? If not I can do so too

Whoops, just updated! I also took @ozankabak's suggestion since there were no objections; thank you!

How is this looking now?

ozankabak reviewed (approved):
LGTM - thank you!

alamb reviewed (approved):
Thank you  @carols10cents and @ozankabak @berkaysynnada  and @crepererum
